hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/DefaultAddressProvider.java

```diff
@@ -54,7 +80,125 @@ public Addresses loadAddresses() {
     }
 
     @Override
-    public Address translate(Address address) {
+    public Address translate(Address address) throws Exception {
+        if (address == null) {
+            return null;
+        }
+
+        if (usePublicAddress) {
+            Address publicAddress = privateToPublic.get(address);
+            if (publicAddress != null) {
+                return publicAddress;
+            }
+        }
         return address;
     }
+
+    @Override
+    public void memberAdded(MembershipEvent membershipEvent) {
+        refresh(membershipEvent.getMembers());
```

Comment:
I think there can be a race between opening a new connection to a new member and updating a `privateToPublic` map because we rely on events. 
`ClientClusterServiceImpl` will set the memberlist first then fire the events. 
As soon as the memberlist is set, `ConnectToAllClusterMembersTask` can kick in and try to open a connection. 
We can potentially call `translate` before we get a chance to update `privateToPublic`. 

Thinking a solution out loud. Can't we solve this on ClientConnectionManager without needing to have a `privatePublic` map ? 
Here instead of calling `translate` https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java#L682
We can just do the following 
```
if(addressProvider instanceof DefaultAddressProvider && usePublicAddress) {
     return member.getAddressMap().get(EndpointQualifier.resolve(ProtocolType.CLIENT, "public"));
}
```
Would that work ?
I need to think more about when to set `usePublicAddress.